relay/plugins/traffic/content-blocker-plugin/content-blocker-plugin.go

```diff
@@ -0,0 +1,296 @@
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	headerBlockers, err := newContentBlockerList(
+		env,
+		"header",
+		excludeHeaderContentVar,
+		maskHeaderContentVar,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(bodyBlockers) == 0 && len(headerBlockers) == 0 {
+		return nil, nil
+	}
+
+	return &contentBlockerPlugin{
+		bodyBlockers:   bodyBlockers,
+		headerBlockers: headerBlockers,
+	}, nil
+}
+
+type contentBlockerPlugin struct {
+	bodyBlockers   []*contentBlocker
+	headerBlockers []*contentBlocker
+}
+
+func (plug contentBlockerPlugin) Name() string {
+	return pluginName
+}
+
+func (plug contentBlockerPlugin) HandleRequest(response http.ResponseWriter, request *http.Request, serviced bool) bool {
+	if serviced := plug.blockHeaderContent(response, request); serviced {
+		return true
+	}
+	if serviced := plug.blockBodyContent(response, request); serviced {
+		return true
+	}
+	return false
+}
+
+func (plug contentBlockerPlugin) blockHeaderContent(response http.ResponseWriter, request *http.Request) bool {
+	if len(plug.headerBlockers) == 0 {
+		return false
+	}
+
+	for _, headerValues := range request.Header {
+		for i, headerValue := range headerValues {
+			processedValue := []byte(headerValue)
+			for _, blocker := range plug.headerBlockers {
+				processedValue = blocker.Block(processedValue)
+			}
+			headerValues[i] = string(processedValue)
+		}
+	}
+
+	return false
+}
+
+func (plug contentBlockerPlugin) blockBodyContent(response http.ResponseWriter, request *http.Request) bool {
+	if len(plug.bodyBlockers) == 0 {
+		return false
+	}
+
+	// Block all websocket connections if we're blocking body content.
+	// TODO(sethf): This is necessary because the current plugin architecture
+	// doesn't give us a hook that would allow us to handle websocket messages,
+	// so we wouldn't be able to actually do any blocking. The safest thing to
+	// do for now is to fail closed. In the short term, this won't do any harm,
+	// because we don't actually need to support websockets, but if that changes
+	// we'll need to revisit this.
+	if len(plug.bodyBlockers) > 0 && request.Header.Get("Upgrade") == "websocket" {
+		logger.Println("Rejecting websocket connection (content blocking is not supported with websockets):", request.URL)
+		http.Error(response, fmt.Sprintf("Blocking unsupported websocket connection: %v", request.URL), 500)
+		return true
+	}
+
+	if request.Body == nil || request.Body == http.NoBody {
+		return false
+	}
+
+	processedBody, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		http.Error(response, fmt.Sprintf("Error reading request body: %s", err), 500)
+		request.Body = http.NoBody
+		return true
+	}
+	initialLength := len(processedBody)
+
+	for _, blocker := range plug.bodyBlockers {
+		processedBody = blocker.Block(processedBody)
+	}
+
+	// If the length of the body has changed, we should update the
+	// Content-Length header too.
+	finalLength := len(processedBody)
+	if finalLength != initialLength {
```

Comment:
Feature request: for the sake of debugging on the fullstory side, could we add a header here (+ I&E side) so that we're aware that _some_ modification had occurred? By the nature of redaction, I know we won't know exactly what happened, but at least we can get a signal that something was modified via relay.

This can be in a follow up.

Reply:
That's easy enough; I can just add it in this PR. I'll push a new version shortly.